### ISSUE TITLE
[HOLD for payment 2024-07-17] [$500] Web – Chat – Blue frame appears around message Reply when open context menu & use keyboard arrow

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#32279]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.6-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Log in with any account.
3. Go to any chat that has a message with Reply
4. Right-click the message
5. Verify that context menu opens
6. Repeat pressing up/down arrows from keyboard

## Expected Result:
Blue frame does not appear around message Reply when open context menu and use up/down arrows from keyboard

## Actual Result:
Blue frame appears around message Reply when open context menu and use up/down arrows from keyboard

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] Windows: Chrome
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/189c89f0-0205-4c26-94f1-b263e084cf8a

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0185e187ff67a1fd61</li>
        <li>Upwork Job ID: 1740328336090607616</li>
        <li>Last Price Increase: 2023-12-28</li>
<li>Automatic offers: </li>
<ul>
<li>situchan | Reviewer | 102947241</li>
<li>bernhardoj | Contributor | 102947244</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @trjExpensify</details>

**Comment by Chetax** [comment]
@getusha Please assign it to me

**Comment by SSRV25** [comment]
I  think this one already sorted.

**Comment by SSRV25** [comment]
It is working good.
https://github.com/Expensify/App/assets/75600326/0dc922c3-9bcc-4701-8e5d-c3cf040a1909

**Comment by SSRV25** [comment]
Contributor details
Your Expensify account email: <sauravmdh25@gmail.com(mailto:sauravmdh25@gmail.com)>
Upwork Profile Link: <https://www.upwork.com/freelancers/~015a4060b1006352d5>

**Comment by SSRV25** [comment]
Contributor details
Your Expensify account email: sauravmdh25@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~015a4060b1006352d5

**Comment by MahmudjonToraqulov** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Web – Chat – Blue frame appears around message Reply when open context menu & use keyboard arrow

### What is the root cause of that problem?
There's no ``` 'box-shadow: none' ```style for this component:
https://github.com/Expensify/App/blob/2e8de3ffd53f40139a927bc60461feadefe88d78/src/pages/home/report/ReportActionItemThread.js#L46

### What changes do you think we should make in order to solve the problem?
We can simply add the style to this comeponent by adding ```style={[ styles.noSelect ]}```
https://github.com/Expensify/App/blob/2e8de3ffd53f40139a927bc60461feadefe88d78/src/pages/home/report/ReportActionItemThread.js#L46

### Result

https://github.com/Expensify/App/assets/91727683/f25d0541-edfe-4998-8564-d6f28cdbd185



### What alternative solutions did you explore? (Optional)
N/A

**Comment by situchan** [comment]
We aren't prioritizing keyboard accessibility issues at the moment. Fine to close

**Comment by getusha** [comment]
unassigned my self as i am not able to get windows device to test this, @situchan offered to take this.

**Comment by situchan** [comment]
I think the issue is that popover is not focused on right click.
i.e. when press Tab key, first focus on composer + rather than reaction icon in popover


https://github.com/Expensify/App/assets/108292595/b4986a0a-0d05-42c3-8602-81a11ee84762

**Comment by situchan** [comment]
Before going ahead, I'd like to confirm with engineer if this is worth fixing as this is accessibility issue
🎀 👀 🎀

**Comment by jasperhuangg** [comment]
I think accessibility issues are important to fix, so let's move forward with a proposal here, @situchan

**Comment by jasperhuangg** [comment]
Still waiting on proposal review

**Comment by situchan** [comment]
Ah sorry was out of my radar as not assigned. Reviewing shortly

**Comment by situchan** [comment]
@tienifr can you share other instances where your solution is applied to fix similar issues?

**Comment by situchan** [comment]
Why can't we fix https://github.com/Expensify/App/issues/32279#issuecomment-1853583261 which is the actual root cause?

**Comment by jasperhuangg** [comment]
I agree with @situchan that we should fix the root cause, which is the focusing the popover first.

**Comment by dragnoir** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Chat – Blue focus (from browser) accessibility frame appears around message Reply when open context menu & use keyboard arrow

### What is the root cause of that problem?
When a user interacts with the context menu in the application using keyboard navigation, an unintended blue focus frame (browser's default focus style) appears around the message reply button. This occurs specifically when the user opens the context menu and navigates using the keyboard arrows.

This issue arises due to the way browsers handle focus on elements. When the context menu is activated (via a right-click on a message reply), the focus remains on the originally clicked element in the DOM, which is the reply thread in our case. As the user navigates through the context menu using the keyboard, the focus style applied by the browser remains visibly attached to the reply thread (button), leading to a confusing user experience.

### What changes do you think we should make in order to solve the problem?

**Adjust Focus Behavior:** Modify the focus behavior when the context menu is triggered. Currently, when a user right-clicks on the message reply, the `onSecondaryInteraction` event is triggered, which calls the `showPopover` function to display the context menu.
https://github.com/Expensify/App/blob/7ae1ec7e74b908e6340bccaf7e42b0345d766711/src/pages/home/report/ReportActionItem.js#L255-L261

**Code Modification:** In the `showPopover` function, introduce a new snippet to manage the focus state (we already use this in other components):
```js
if (DomUtils.getActiveElement()) {
  DomUtils.getActiveElement().blur();
}
```
This code ensures that the focus is removed from the message reply DOM element when the context menu is activated, thereby preventing the display of the browser's default focus style on the reply button.

### What alternative solutions did you explore? (Optional)

**Third-Party Library Integration:** One alternative could be to integrate a third-party library specifically designed to manage focus states and facilitate proper tab navigation. However, this approach may be overly complex for addressing a relatively minor issue. It would also necessitate extensive modifications to other parts of the application's tab navigation system.

**Comment by situchan** [comment]
reviewing new proposals

**Comment by situchan** [comment]
@tienifr please do regression test with your new solution as it affects all popovers throughout the app.
`focus-trap-react` module is being added in #32800 which is 3rd attempt (because of 2 regressions) of fixing #15631.

**Comment by bernhardoj** [comment]
Just FYI, we already have a proposed solution [here](https://github.com/Expensify/App/issues/31756#issuecomment-1824095948) of wrapping the popover with focus trap which is currently on hold for https://github.com/Expensify/App/issues/15631

**Comment by situchan** [comment]
Same

**Comment by situchan** [comment]
Still holding for [focus trap v3](https://github.com/Expensify/App/pull/32800)

**Comment by situchan** [comment]
Same

**Comment by situchan** [comment]
still holding

**Comment by situchan** [comment]
Based on https://github.com/Expensify/App/pull/32800#issuecomment-1924310126, looks like holding PR is going to be closed.

**Comment by jasperhuangg** [comment]
@roryabraham looks like we can close out the PR, does that mean the issue can also be closed out?

**Comment by situchan** [comment]
The linked issue fixed during navigation refactor but this bug still exists though

<img width="1398" alt="bug" src="https://github.com/Expensify/App/assets/108292595/f8a646a9-f933-43db-ae8d-7fbcae563534">

**Comment by jasperhuangg** [comment]
@situchan sounds good, let's open this back up for proposals

**Comment by dragnoir** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Blue frame on wrong focused element due to right click.

### What is the root cause of that problem?

This issue is not happening only on the element the QA mentioned on the issue. This happens with alot of element inside the report screen. Below some:

- icon on comment:
![image](https://github.com/Expensify/App/assets/12425932/ffb3d1c2-b539-4ac4-8c86-28f0d27c5635)

- profile picture
![image](https://github.com/Expensify/App/assets/12425932/537bdfb7-0fec-4849-8ad4-decbf6f2f1a9)

- .. and more

I  tested  `<FocusTrap>` and it's not solving the issue.
I also tested `withoutFocusOnSecondaryInteraction` and it's removing the focus from the selected element but when you click on TAB from keyboard, the focus goes to other wrong element on the report screen.

The cause of the issue is that when the Report Action Context Menu is opened, the focus is not inside the component, it still outside of it.

### What changes do you think we should make in order to solve the problem?

We need to set the focus inside the `PopoverReportActionContextMenu` so when we use keyboard keys, we navigate to the elments inside the component.

below a part of the main code change.
We need to set focus on the 1st element from the component here
https://github.com/Expensify/App/blob/9a598e0437cecb9bc23a57d7f15b59d496b69631/src/components/Reactions/QuickEmojiReactions/BaseQuickEmojiReactions.tsx#L34-L38

```diff
- {CONST.QUICK_REACTIONS.map((emoji: Emoji) => (
+ {CONST.QUICK_REACTIONS.map((emoji: Emoji, index: number) => (
  <Tooltip
   text={`:${EmojiUtils.getLocalizedEmojiName(emoji.name, preferredLocale)}:`}
   key={emoji.name}
  >
-  <View>
+  <View  ref={index === 0 ? firstEmojiRef : null}>
    <EmojiReactionBubble
      emojiCodes={[EmojiUtils.getPreferredEmojiCode(emoji, preferredSkinTone)]}
      isContextMenu
      onPress={Session.checkIfActionIsAllowed(() =>  onEmojiSelected(emoji, emojiReactions))}
    />
  </View>
  </Tooltip>
))}
```

Once we get the ref of the 1st element, we need to set focus to it:
```js
useEffect(() => {
  // Set focus to the first emoji element
  if (!firstEmojiRef.current) {
   return;
  }
  const  buttonElement  =  firstEmojiRef.current.querySelector('button');
  if (buttonElement) {
   buttonElement.focus();
  }
}, []);
```

With the update below, once the Popover is opened,  we now have the focus set to the element we want.

POC (I let the focus with the blue border for the demo video to show how it's working):

https://github.com/Expensify/App/assets/12425932/6d56e724-597e-470c-a2d2-cc04c469099b

The DEMO was a quick code edit, deep details can be handled throw the PR.

**Comment by dragnoir** [comment]
New proposal [here](https://github.com/Expensify/App/issues/32279#issuecomment-1946110514)

**Comment by situchan** [comment]
@dragnoir thanks for the proposal but it looks like workaround. I don't like the idea of getting ref of first element.
Looking for more general solution.

**Comment by MahmudjonToraqulov** [comment]
Hey @situchan. What about my [solution](https://github.com/Expensify/App/issues/32279#issuecomment-1835873259). You didn't give a review for it so that I update it. Could you check it and give a review pls?

**Comment by situchan** [comment]
@MahmudjonToraqulov please try to find general solution. Not only happening on thread item



https://github.com/Expensify/App/assets/108292595/a202b5f7-2c09-427a-b63f-3fdc65193f9f

**Comment by dragnoir** [comment]
@situchan pls check my [new alternative solutions](https://github.com/Expensify/App/issues/32279#issuecomment-1946110514)

**Comment by trjExpensify** [comment]
So @mountiny this is keyboard/tab related. Are we going to hold these and move them to a more broader project?

**Comment by situchan** [comment]
I think we can continue.
I asked [here](https://github.com/Expensify/App/issues/32279#issuecomment-1853585953) and here's the [answer](https://github.com/Expensify/App/issues/32279#issuecomment-1858525238)

**Comment by dragnoir** [comment]
@situchan this is a video that shows the part I mentioned in my alternative solution (LHN) and we should follow

https://github.com/Expensify/App/assets/12425932/23beb8c7-34ae-4f05-98a1-2bd1c1d5def1

**Comment by trjExpensify** [comment]
>I think we can continue.
I asked https://github.com/Expensify/App/issues/32279#issuecomment-1853585953 and here's the https://github.com/Expensify/App/issues/32279#issuecomment-1858525238

Okay, but thinking about the wider context. We haven't put an emphasis on accessibility and particularly keyboard/tab navigation throughout the app. There's a ton of places where we have "bugs" in this regard. Equally, we don't have anything in the checklist or guidelines for developing to ensure we're always accounting for this navigation -- so we'll forever have "new bugs" as a result of not officially supporting it and putting a framework/guidelines around it. 

Fixing these issues "one off" has diminishing returns, so a better approach to truly support this consistently throughout the app would to tackle it holistically and methodically. @mountiny @puneetlath @iwiznia and I discussed that a bit internally, and I've moved the output of that to [here](https://github.com/Expensify/App/issues/36476#issuecomment-1959232906) to see if we can get the ball rolling to put this to bed once and for all.

**Comment by situchan** [comment]
@trjExpensify thanks for the summary. So you suggest to hold this issue for https://github.com/Expensify/App/issues/36476, right?

**Comment by mountiny** [comment]
I agree with the hold

**Comment by jasperhuangg** [comment]
Still on HOLD

**Comment by situchan** [comment]
Same

**Comment by trjExpensify** [comment]
Held on #36476 still.

**Comment by trjExpensify** [comment]
Samesies.

**Comment by jasperhuangg** [comment]
Held on https://github.com/Expensify/App/issues/36476 not overdue

**Comment by trjExpensify** [comment]
Same!

**Comment by jasperhuangg** [comment]
Still held on https://github.com/Expensify/App/issues/36476

**Comment by trjExpensify** [comment]
Still held, Melv. @jnowakow is picking up from Adam on https://github.com/Expensify/App/issues/36476#issuecomment-2095496118

**Comment by trjExpensify** [comment]
Same, Melv.

**Comment by situchan** [comment]
Same ^

**Comment by trjExpensify** [comment]
No change.

**Comment by situchan** [comment]
https://github.com/Expensify/App/pull/39520 which fixes #36476 was deployed to prod on Jun 14.

**Comment by situchan** [comment]
Bug is still reproducible

![image](https://github.com/Expensify/App/assets/108292595/a4708504-4936-4e67-b436-5453dd8fff76)

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Blue outline appears around the reply button when pressing arrow up key while context menu is open.

### What is the root cause of that problem?
When we right click the reply/replies button, the button is focused. If we press any key, the blue outline will appear around the focused element. It's supposed to be fixed by a focus trap PR so the context menu will be focused when opened, but then we decided to not focus the context menu/any modal initially in this [PR](https://github.com/Expensify/App/pull/43743).

However, if we use the arrow down key, then the blue outline around the reply/replies doesn't appear. Why? It's because pressing the arrow down key will highlight and focus (thanks to useSyncFocus) the first context menu item, but if we use the arrow up key, none of the context menu items are highlighted and focused.

Previously, if we press the arrow up key, the last item will be highlighted and focused, but in this [commit](https://github.com/Expensify/App/pull/39520/commits/97408438b6830a544d9e5fb504f9e32dfa98f0a6), we disable the cyclic traversal of the context menu arrow key.
https://github.com/Expensify/App/blob/b33542383ac2cb8ea1e922aa12dd2265efccd0ab/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L167

### What changes do you think we should make in order to solve the problem?

Enable the cyclic traversal back by removing this code.
https://github.com/Expensify/App/blob/b33542383ac2cb8ea1e922aa12dd2265efccd0ab/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L167


https://github.com/Expensify/App/assets/50919443/eee2f234-62e3-4dcf-a9fe-fa34df35a4e7

**Comment by situchan** [comment]
@bernhardoj thanks for following this issue.

> Previously, if we press the arrow up key, the last item will be highlighted and focused, but in this [commit](https://github.com/Expensify/App/pull/39520/commits/97408438b6830a544d9e5fb504f9e32dfa98f0a6), we disable the cyclic traversal of the context menu arrow key.

If it was intentionally disabled, then won't enabling back cause regression?

**Comment by trjExpensify** [comment]
Bump, @jnowakow. Looking for input [on this](https://github.com/Expensify/App/issues/32279#issuecomment-2175051476).

**Comment by jasperhuangg** [comment]
Still waiting on a response from @jnowakow

**Comment by jnowakow** [comment]
Hello, sorry that you had to wait so long for me. 
The cyclic travers was disabled because it would break the tab navigation. When last element was focus clicking tab would move focus to first emoji but also highlight the first element of list which would be confusing. Without cyclic travel there's no such problem

**Comment by bernhardoj** [comment]
I see. Thanks for the explanation, but I tested it out and it works fine. I think the issue doesn't persist anymore.

(cyclic traversal enabled)

https://github.com/Expensify/App/assets/50919443/e8517dac-c059-40fa-a2d1-c46f05b228e4

**Comment by trjExpensify** [comment]
Okay, so what are the next steps here?

**Comment by bernhardoj** [comment]
I think we need @situchan approval

**Comment by situchan** [comment]
@bernhardoj's [proposal](https://github.com/Expensify/App/issues/32279#issuecomment-2175047572) looks good to me.
🎀👀🎀 C+ reviewed

**Comment by bernhardoj** [comment]
PR is ready

cc: @situchan

**Comment by bernhardoj** [comment]
I'll request in ND once payment is due.

**Comment by bernhardoj** [comment]
The payment should be due [tomorrow](https://github.com/Expensify/App/issues/32279#issuecomment-2220202116) (07-17).

**Comment by bernhardoj** [comment]
Requested in ND.

**Comment by trjExpensify** [comment]
Yup, awaiting the checklist!

**Comment by situchan** [comment]
Not able to find offending PR. I think this bug existed from the beginning.
As this is accessibility issue, I don't think it's worth adding regression test.

**Comment by trjExpensify** [comment]
Alright, payment summary as follows: 

- $500 to @bernhardoj for the fix (go ahead and request)
- $500 to @situchan for the C+ (paid via upwork!)


Internal note: The Upwork contract has a [start date](https://www.upwork.com/nx/wm/offer/102947244) of 1st July, and Bernhard was eligible from 27th June according to this [SO)](https://stackoverflowteams.com/c/expensify/questions/16949).

Closing!

**Comment by JmillsExpensify** [comment]
$500 approved for @bernhardoj
